stdlib/public/Concurrency/TaskCancellation.swift

```diff
@@ -0,0 +1,169 @@
+//// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+////
+////===----------------------------------------------------------------------===//
+
+import Swift
+@_implementationOnly import _SwiftConcurrencyShims
+
+// ==== Task Cancellation ------------------------------------------------------
+
+extension Task {
+
+  /// Returns `true` if the task is cancelled, and should stop executing.
+  ///
+  /// - SeeAlso: `Task.checkCancellation`
+  public static func isCancelled() async -> Bool {
+    // let task = __getTask() // TODO: pending internal API to get tasks
+    // task.isCancelled || task.deadline.isOverdue
+    fatalError("\(#function) not implemented yet.")
+  }
+
+  /// Check if the task is cancelled and throw an `CancellationError` if it was.
+  ///
+  /// It is intentional that no information is passed to the task about why it
+  /// was cancelled. A task may be cancelled for many reasons, and additional
+  /// reasons may accrue / after the initial cancellation (for example, if the
+  /// task fails to immediately exit, it may pass a deadline).
+  ///
+  /// The goal of cancellation is to allow tasks to be cancelled in a
+  /// lightweight way, not to be a secondary method of inter-task communication.
+  public static func checkCancellation() async throws {
+    if await Task.isCancelled() {
+      throw CancellationError()
+    }
+  }
+
+  /// Execute an operation with cancellation handler which will immediately be
+  /// invoked if the current task is cancelled.
+  ///
+  /// This differs from the operation cooperatively checking for cancellation
+  /// and reacting to it in that the cancellation handler is _always_ and
+  /// _immediately_ invoked when the task is cancelled. For example, even if the
+  /// operation is running code which never checks for cancellation, a cancellation
+  /// handler still would run and give us a chance to run some cleanup code.
+  ///
+  /// Does not check for cancellation, and always executes the passed `operation`.
+  public static func withCancellationHandler<T>(
+    handler: /* @concurrent */ () -> (),
+    operation: () async throws -> T
+  ) async throws -> T {
+      fatalError("\(#function) not implemented yet.")
+  }
+
+  /// The default cancellation thrown when a task is cancelled.
+  ///
+  /// This error is also thrown automatically by `Task.checkCancellation()`,
+  /// if the current task has been cancelled.
+  public struct CancellationError: Error {
+    // no extra information, cancellation is intended to be light-weight
+  }
+
+}
+
+// ==== Task Deadlines ---------------------------------------------------------
+
+extension Task {
+
+  /// Returns the earliest deadline set on the current task.
+  ///
+  /// If no deadline was set for the task the `Deadline.distantFuture` is returned,
+  /// as it is effective in conveying that there still is time remaining and the
+  /// deadline is not overdue yet.
+  public static func currentDeadline() -> Deadline {
```

Comment:
Should this be an async function as other APIs?
```suggestion
  public static func currentDeadline() async -> Deadline {
```

Reply:
Yes, bringing the PR up to date now.